File: python/175.py
"""Project Euler Problem 175: Fractions involving the number of different ways a number can be expressed as a sum of powers of 2."""

from typing import List
import math

P_TARGET = 123_456_789
Q_TARGET = 987_654_321

POWERS = [1 << k for k in range(61)]


def continued_fraction(num: int, den: int) -> List[int]:
    """Simple continued fraction of num/den."""
    cf: List[int] = []
    while den != 0:
        a = num // den
        cf.append(a)
        num, den = den, num - a * den
    return cf


def ensure_parity(cf: List[int], desired_parity: int) -> List[int]:
    """Adjust the continued fraction so its length parity matches desired."""
    cf = cf.copy()
    while len(cf) % 2 != desired_parity:
        if cf[-1] > 1:
            cf[-1] -= 1
            cf.append(1)
        elif len(cf) == 1:
            cf[0] += 1
        else:
            cf[-2] += 1
            cf.pop()
    return cf


def runs_from_fraction(p: int, q: int) -> List[int]:
    """Convert fraction to run lengths for the shortened binary expansion."""
    if p <= 0 or q <= 0:
        raise ValueError("p and q must be positive")

    g = math.gcd(p, q)
    p //= g
    q //= g
    if p == q:
        return [1]

    if p < q:
        cf = continued_fraction(q, p)
        cf = ensure_parity(cf, 1)  # expect odd number of runs when ratio < 1
    else:
        cf = continued_fraction(p, q)
        cf = ensure_parity(cf, 0)  # expect even number of runs when ratio >= 1

    cf.reverse()
    return cf


def main() -> str:
    """Main function."""
    runs = runs_from_fraction(P_TARGET, Q_TARGET)
    return ",".join(str(r) for r in runs)


if __name__ == "__main__":
    print(main())

File: c/175.c
/*
 * Project Euler Problem 175: Fractions involving the number of different
 * ways a number can be expressed as a sum of powers of 2.
 *
 * Convert p/q = 123456789/987654321 to a continued fraction,
 * adjust parity, reverse, and output as comma-separated run lengths.
 */
#include <stdio.h>

typedef long long i64;

static i64 gcd(i64 a, i64 b) {
    while (b) { i64 t = b; b = a % b; a = t; }
    return a;
}

int main(void) {
    i64 p = 123456789LL;
    i64 q = 987654321LL;

    i64 g = gcd(p, q);
    p /= g;
    q /= g;

    /* Continued fraction of the ratio */
    i64 cf[200];
    int cf_len = 0;

    int swap = 0;
    i64 num, den;
    int desired_parity;

    if (p < q) {
        num = q; den = p;
        desired_parity = 1; /* odd length */
        swap = 1;
    } else {
        num = p; den = q;
        desired_parity = 0; /* even length */
    }

    i64 a_n = num, b_n = den;
    while (b_n != 0) {
        i64 a = a_n / b_n;
        cf[cf_len++] = a;
        i64 tmp = b_n;
        b_n = a_n - a * b_n;
        a_n = tmp;
    }

    /* Adjust parity */
    while (cf_len % 2 != desired_parity) {
        if (cf[cf_len - 1] > 1) {
            cf[cf_len - 1]--;
            cf[cf_len++] = 1;
        } else if (cf_len == 1) {
            cf[0]++;
        } else {
            cf[cf_len - 2]++;
            cf_len--;
        }
    }

    /* Reverse */
    for (int i = 0; i < cf_len / 2; i++) {
        i64 tmp = cf[i];
        cf[i] = cf[cf_len - 1 - i];
        cf[cf_len - 1 - i] = tmp;
    }

    /* Print comma-separated */
    for (int i = 0; i < cf_len; i++) {
        if (i > 0) printf(",");
        printf("%lld", cf[i]);
    }
    printf("\n");

    return 0;
}
